### ISSUE TITLE
Make composite_enum generalized

### ISSUE THREAD

**Issue by xlc** [paritytech/substrate#13821]
Followup of #13722 

Right now it now only accepts FreezeReason, HoldReason, LockId and SlashReason as identifiers for the enum.

Should generalize the implementation to make it possible to extend identifier types.

**Comment by bkchr** [comment]
IMO calling it `composite_enum` and then matching on the name was mistake. There should be an attribute for every type we want to combine in the runtime. Random identifiers are not a great idea, because people otherwise will do simple spelling mistakes and debug for hours why their pallet enum isn't part of the composite in the runtime. However, we could create some kind of `composite_enum` macro that can generate a custom attribute that will then composite in the runtime.

**Comment by gilescope** [comment]
composite_enum without something like explicit pallet_index (i.e. explicit discriminators) seems like a footgun to me.

**Comment by bkchr** [comment]
Not sure what you are talking about, on the runtime side they reuse the pallet index like all the other things.

**Comment by ggwpez** [comment]
>IMO calling it composite_enum and then matching on the name was mistake. There should be an attribute for every type we want to combine in the runtime. Random identifiers are not a great idea, because people otherwise will do simple spelling mistakes and debug for hours why their pallet enum isn't part of the composite in the runtime. 

Do you mean the current `master` is problematic? We do have an UI test that checks for typo reports:  
https://github.com/paritytech/substrate/blob/057f2afd686dceb48c08d8a31f35b338bb5992d3/frame/support/test/tests/pallet_ui/composite_enum_unsupported_identifier.stderr#L1

**Comment by bkchr** [comment]
> Do you mean the current `master` is problematic?

Yes. Yeah, the UI test is fine and a good idea. However, discoverability of this is hard IMO. We also have for all the other things separate macros.
